File: go/memory/churn.go
package main

import (
	"github.com/mohamedmenasy/go-vs-rust-benchmarks/go/internal/common"
	"github.com/mohamedmenasy/go-vs-rust-benchmarks/go/internal/harness"
)

// churn: a live set of `bytes` worth of boxed objects stays reachable (built
// in untimed setup) while each Run replaces `replace` randomly chosen slots
// with fresh objects. In Go the old objects become garbage and every GC
// cycle must mark the whole live set; in Rust the assignment frees the old
// object immediately. Object content depends only on the slot index, so the
// live set is identical after every Run and digests are reproducible.
type churnW struct {
	n, k int
	seed uint64
	live []*obj
}

func (w *churnW) Run() uint64 {
	r := common.NewSplitMix64(w.seed)
	n := uint64(w.n)
	var acc uint64
	for j := 0; j < w.k; j++ {
		i := r.Below(n)
		o := &obj{}
		fillObj(o, i)
		w.live[i] = o
		acc += o.a ^ o.h
	}
	return acc
}

func (w *churnW) Check() uint64 {
	var acc uint64
	for _, o := range w.live {
		acc += o.a ^ o.h
	}
	return acc
}

func (w *churnW) Teardown() {
	w.live = nil
	release()
}

func (w *churnW) Extra() map[string]any {
	return map[string]any{"live_objects": w.n, "replacements": w.k, "live_payload_bytes": w.n * objSize}
}

func setupChurn(p harness.Params) (harness.Instance, harness.Work, error) {
	n := int(p.Int("bytes", 100<<20)) / objSize
	k := int(p.Int("replace_pct", 100)) * n / 100
	live := make([]*obj, n)
	for i := range live {
		o := &obj{}
		fillObj(o, uint64(i))
		live[i] = o
	}
	return &churnW{n: n, k: k, seed: uint64(p.Int("seed", 7)), live: live},
		harness.Work{Unit: "allocations", PerRun: float64(k)}, nil
}

File: rust/memory/src/churn.rs
//! Live-set churn (mirror of go/memory/churn.go): a live set of `bytes`
//! worth of boxed objects stays reachable while each run replaces
//! `replace_pct`% as many randomly chosen slots with fresh objects. The
//! assignment frees the replaced object immediately.

use bench_common::serde_json::{Value, json};
use bench_common::{Instance, Params, SplitMix64, Work};

use crate::{OBJ_SIZE, Obj, release};

struct Churn {
    n: usize,
    k: usize,
    seed: u64,
    live: Vec<Box<Obj>>,
}

impl Instance for Churn {
    fn run(&mut self) -> u64 {
        let mut r = SplitMix64::new(self.seed);
        let n = self.n as u64;
        let mut acc = 0u64;
        for _ in 0..self.k {
            let i = r.below(n);
            let o = Box::new(Obj::filled(i));
            acc = acc.wrapping_add(o.a ^ o.h);
            self.live[i as usize] = o;
        }
        acc
    }
    fn has_check(&self) -> bool {
        true
    }
    fn check(&mut self) -> u64 {
        self.live.iter().fold(0u64, |acc, o| acc.wrapping_add(o.a ^ o.h))
    }
    fn has_teardown(&self) -> bool {
        true
    }
    fn teardown(&mut self) {
        self.live = Vec::new();
        release();
    }
    fn extra(&mut self) -> Option<Value> {
        Some(json!({"live_objects": self.n, "replacements": self.k, "live_payload_bytes": self.n * OBJ_SIZE}))
    }
}

pub fn setup(p: &Params) -> Result<(Box<dyn Instance>, Work), String> {
    let n = p.int("bytes", 100 << 20) as usize / OBJ_SIZE;
    let k = p.int("replace_pct", 100) as usize * n / 100;
    let live: Vec<Box<Obj>> = (0..n).map(|i| Box::new(Obj::filled(i as u64))).collect();
    Ok((
        Box::new(Churn { n, k, seed: p.int("seed", 7) as u64, live }),
        Work { unit: "allocations", per_run: k as f64, input_bytes: 0 },
    ))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::testutil::{golden_memory, params, run_twice};
    use bench_common::golden_hex;

    #[test]
    fn matches_golden() {
        let c = &golden_memory()["churn"];
        let bytes = (c["n"].as_u64().unwrap() as usize * OBJ_SIZE).to_string();
        let (mut inst, _) = setup(&params(&[
            ("bytes", bytes),
            ("replace_pct", "100".to_string()),
            ("seed", c["seed"].to_string()),
        ]))
        .unwrap();
        assert_eq!(run_twice(&mut inst), golden_hex(&c["run"]));
        assert_eq!(inst.check(), golden_hex(&c["check"]));
    }
}
